reana_server/complexity.py

```diff
@@ -0,0 +1,360 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of REANA.
+# Copyright (C) 2021 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""REANA workflow complexity estimation."""
+
+import os
+from functools import reduce
+
+from reana_commons.job_utils import kubernetes_memory_to_bytes
+from reana_commons.config import MQ_MAX_PRIORITY
+
+from reana_server.status import NodesStatus
+from reana_server.config import REANA_KUBERNETES_DEFAULT_JOBS_MEMORY_LIMIT
+
+
+def get_workflow_min_job_memory(complexity):
+    """Return minimal job memory from workflow complexity.
+
+    :param complexity: workflow complexity list which consists of number of initial jobs and the memory in bytes they require. (e.g. [(8, 1073741824), (5, 2147483648)])
+    :return: minimal job memory (e.g. 1073741824)
+    """
+    if not complexity:
+        return 0
+    return min(complexity, key=lambda x: x[1])[1]
+
+
+def calculate_workflow_priority(complexity):
+    """Calculate workflow priority based on its complexity."""
+    if not complexity:
+        return 0
+    wf_memory = int(reduce(lambda sum, item: sum + item[0] * item[1], complexity, 0))
+    total_cluster_memory = NodesStatus().get_total_memory()
+    if wf_memory > total_cluster_memory:
+        return 1
```

Comment:
Why 1 here? Is this to force a reschedule?

Reply:
This is a protection not to exceed over `MQ_MAX_PRIORITY` (which is 100 in our case). If a workflow requires more memory than a cluster can offer, then by our formula `int((1 - wf_memory / total_cluster_memory) * MQ_MAX_PRIORITY)` the priority could be set higher than `MQ_MAX_PRIORITY` and break the scheduler. So in this case I set it to 1, since it's a minimum priority there could be (it could be also set to 0, but that would more mean that there is no priority for this workflow)